Permissions error with config mounted into MySQL container

```
11:42:05.293 [Test worker] ERROR 🐳 [mysql:5.7.22] - Container log output (if any) will follow:
11:42:05.304 [tc-okhttp-stream-474827747] INFO 🐳 [mysql:5.7.22] - STDERR: 
11:42:05.305 [tc-okhttp-stream-474827747] INFO 🐳 [mysql:5.7.22] - STDERR: ERROR: mysqld failed while attempting to check config
11:42:05.305 [tc-okhttp-stream-474827747] INFO 🐳 [mysql:5.7.22] - STDERR: command was: "mysqld --verbose --help"
11:42:05.305 [tc-okhttp-stream-474827747] INFO 🐳 [mysql:5.7.22] - STDERR: 
11:42:05.306 [tc-okhttp-stream-474827747] INFO 🐳 [mysql:5.7.22] - STDERR: mysqld: Can't read dir of '/etc/mysql/conf.d/' (Errcode: 13 - Permission denied)
11:42:05.306 [tc-okhttp-stream-474827747] INFO 🐳 [mysql:5.7.22] - STDERR: mysqld: [ERROR] Fatal error in defaults handling. Program aborted!
```

We see this in the log (look on the null):
`
Cmd: com.github.dockerjava.core.command.CopyArchiveToContainerCmdImpl@2c8c6bb[cp ,<null>, ,6ca1ea3ce4b2582536fac69ceeb991670930b1ac65a6b2206c1ab88f3c990b65,:,/etc/mysql/]
11:41:45.064 [Test worker] INFO 🐳 [mysql:5.7.22] - Starting container with ID: 6ca1ea3ce4b2582536fac69ceeb991670930b1ac65a6b2206c1ab88f3c990b65
`

If we override the MySQLContainer class and just rmove this line from the configure function:
`
optionallyMapResourceParameterAsVolume(MY_CNF_CONFIG_OVERRIDE_PARAM_NAME, "/etc/mysql/conf.d", "mysql-default-conf");
`

Everything is working. Looks like it is copying invalid file to the container

Same Issue here 

```
build	24-May-2019 09:21:02	2019-05-24 09:21:02.234 ERROR 605 --- [           main] ?.7.23] : Log output from the failed container:
build	24-May-2019 09:21:02	ERROR: mysqld failed while attempting to check config
build	24-May-2019 09:21:02	
build	24-May-2019 09:21:02	command was: "mysqld --verbose --help"
build	24-May-2019 09:21:02	
build	24-May-2019 09:21:02	mysqld: Can't read dir of '/etc/mysql/conf.d/' (Errcode: 13 - Permission denied)
build	24-May-2019 09:21:02	mysqld: [ERROR] Fatal error in defaults handling. Program aborted!
```

--- a/modules/jdbc/src/main/java/org/testcontainers/containers/JdbcDatabaseContainer.java
+++ b/modules/jdbc/src/main/java/org/testcontainers/containers/JdbcDatabaseContainer.java
@@ -5,6 +5,7 @@ import lombok.NonNull;
 import lombok.SneakyThrows;
 import org.apache.commons.lang3.StringUtils;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 import org.testcontainers.containers.traits.LinkableContainer;
 import org.testcontainers.delegate.DatabaseDelegate;
 import org.testcontainers.ext.ScriptUtils;
@@ -302,15 +303,25 @@ public abstract class JdbcDatabaseContainer<SELF extends JdbcDatabaseContainer<S
         return urlParameters;
     }
 
+    @Deprecated
     protected void optionallyMapResourceParameterAsVolume(
         @NotNull String paramName,
         @NotNull String pathNameInContainer,
         @NotNull String defaultResource
     ) {
+        optionallyMapResourceParameterAsVolume(paramName, pathNameInContainer, defaultResource, null);
+    }
+
+    protected void optionallyMapResourceParameterAsVolume(
+        @NotNull String paramName,
+        @NotNull String pathNameInContainer,
+        @NotNull String defaultResource,
+        @Nullable Integer fileMode
+    ) {
         String resourceName = parameters.getOrDefault(paramName, defaultResource);
 
         if (resourceName != null) {
-            final MountableFile mountableFile = MountableFile.forClasspathResource(resourceName);
+            final MountableFile mountableFile = MountableFile.forClasspathResource(resourceName, fileMode);
             withCopyFileToContainer(mountableFile, pathNameInContainer);
         }
     }

--- a/modules/mariadb/src/main/java/org/testcontainers/containers/MariaDBContainer.java
+++ b/modules/mariadb/src/main/java/org/testcontainers/containers/MariaDBContainer.java
@@ -1,6 +1,7 @@
 package org.testcontainers.containers;
 
 import com.google.common.collect.Sets;
+import org.testcontainers.images.builder.Transferable;
 import org.testcontainers.utility.DockerImageName;
 
 import java.util.Set;
@@ -67,7 +68,8 @@ public class MariaDBContainer<SELF extends MariaDBContainer<SELF>> extends JdbcD
         optionallyMapResourceParameterAsVolume(
             MY_CNF_CONFIG_OVERRIDE_PARAM_NAME,
             "/etc/mysql/conf.d",
-            "mariadb-default-conf"
+            "mariadb-default-conf",
+            Transferable.DEFAULT_DIR_MODE
         );
 
         addEnv("MYSQL_DATABASE", databaseName);

--- a/modules/mariadb/src/test/java/org/testcontainers/junit/mariadb/SimpleMariaDBTest.java
+++ b/modules/mariadb/src/test/java/org/testcontainers/junit/mariadb/SimpleMariaDBTest.java
@@ -6,10 +6,19 @@ import org.testcontainers.MariaDBTestImages;
 import org.testcontainers.containers.MariaDBContainer;
 import org.testcontainers.db.AbstractContainerDatabaseTest;
 
+import java.io.File;
+import java.net.URL;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.attribute.PosixFilePermission;
 import java.sql.ResultSet;
 import java.sql.SQLException;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assumptions.assumeThat;
 import static org.junit.Assume.assumeFalse;
 
 public class SimpleMariaDBTest extends AbstractContainerDatabaseTest {
@@ -56,10 +65,7 @@ public class SimpleMariaDBTest extends AbstractContainerDatabaseTest {
         ) {
             mariadbCustomConfig.start();
 
-            ResultSet resultSet = performQuery(mariadbCustomConfig, "SELECT @@GLOBAL.innodb_file_format");
-            String result = resultSet.getString(1);
-
-            assertThat(result).as("The InnoDB file format has been set by the ini file content").isEqualTo("Barracuda");
+            assertThatCustomIniFileWasUsed(mariadbCustomConfig);
         }
     }
 
@@ -92,6 +98,44 @@ public class SimpleMariaDBTest extends AbstractContainerDatabaseTest {
             assertThat(jdbcUrl).contains("connectTimeout=40000");
         } finally {
             mariaDBContainer.stop();
+        }
+    }
+
+    @Test
+    public void testWithOnlyUserReadableCustomIniFile() throws Exception {
+        assumeThat(FileSystems.getDefault().supportedFileAttributeViews().contains("posix")).isTrue();
+
+        try (
+            MariaDBContainer<?> mariadbCustomConfig = new MariaDBContainer<>(
+                MariaDBTestImages.MARIADB_IMAGE.withTag("10.1.16")
+            )
+                .withConfigurationOverride("somepath/mariadb_conf_override")
+        ) {
+            URL resource = this.getClass().getClassLoader().getResource("somepath/mariadb_conf_override");
+
+            File file = new File(resource.toURI());
+            assertThat(file.isDirectory()).isTrue();
+
+            Set<PosixFilePermission> permissions = new HashSet<>(
+                Arrays.asList(
+                    PosixFilePermission.OWNER_READ,
+                    PosixFilePermission.OWNER_WRITE,
+                    PosixFilePermission.OWNER_EXECUTE
+                )
+            );
+
+            Files.setPosixFilePermissions(file.toPath(), permissions);
+
+            mariadbCustomConfig.start();
+
+            assertThatCustomIniFileWasUsed(mariadbCustomConfig);
+        }
+    }
+
+    private void assertThatCustomIniFileWasUsed(MariaDBContainer<?> mariadb) throws SQLException {
+        try (ResultSet resultSet = performQuery(mariadb, "SELECT @@GLOBAL.innodb_file_format")) {
+            String result = resultSet.getString(1);
+            assertThat(result).as("The InnoDB file format has been set by the ini file content").isEqualTo("Barracuda");
         }
     }
 }

--- a/modules/mysql/src/main/java/org/testcontainers/containers/MySQLContainer.java
+++ b/modules/mysql/src/main/java/org/testcontainers/containers/MySQLContainer.java
@@ -1,6 +1,7 @@
 package org.testcontainers.containers;
 
 import org.jetbrains.annotations.NotNull;
+import org.testcontainers.images.builder.Transferable;
 import org.testcontainers.utility.DockerImageName;
 
 import java.util.Set;
@@ -71,7 +72,8 @@ public class MySQLContainer<SELF extends MySQLContainer<SELF>> extends JdbcDatab
         optionallyMapResourceParameterAsVolume(
             MY_CNF_CONFIG_OVERRIDE_PARAM_NAME,
             "/etc/mysql/conf.d",
-            "mysql-default-conf"
+            "mysql-default-conf",
+            Transferable.DEFAULT_DIR_MODE
         );
 
         addEnv("MYSQL_DATABASE", databaseName);

--- a/modules/mysql/src/test/java/org/testcontainers/junit/mysql/SimpleMySQLTest.java
+++ b/modules/mysql/src/test/java/org/testcontainers/junit/mysql/SimpleMySQLTest.java
@@ -9,6 +9,11 @@ import org.testcontainers.containers.MySQLContainer;
 import org.testcontainers.containers.output.Slf4jLogConsumer;
 import org.testcontainers.db.AbstractContainerDatabaseTest;
 
+import java.io.File;
+import java.net.URL;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.attribute.PosixFilePermission;
 import java.sql.Connection;
 import java.sql.ResultSet;
 import java.sql.SQLException;
@@ -17,9 +22,13 @@ import java.time.LocalDateTime;
 import java.time.ZoneId;
 import java.time.ZonedDateTime;
 import java.time.temporal.ChronoUnit;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.fail;
+import static org.assertj.core.api.Assumptions.assumeThat;
 
 public class SimpleMySQLTest extends AbstractContainerDatabaseTest {
 
@@ -84,10 +93,7 @@ public class SimpleMySQLTest extends AbstractContainerDatabaseTest {
         ) {
             mysqlCustomConfig.start();
 
-            ResultSet resultSet = performQuery(mysqlCustomConfig, "SELECT @@GLOBAL.innodb_file_format");
-            String result = resultSet.getString(1);
-
-            assertThat(result).as("The InnoDB file format has been set by the ini file content").isEqualTo("Barracuda");
+            assertThatCustomIniFileWasUsed(mysqlCustomConfig);
         }
     }
 
@@ -234,8 +240,43 @@ public class SimpleMySQLTest extends AbstractContainerDatabaseTest {
         }
     }
 
+    @Test
+    public void testWithOnlyUserReadableCustomIniFile() throws Exception {
+        assumeThat(FileSystems.getDefault().supportedFileAttributeViews().contains("posix")).isTrue();
+        try (
+            MySQLContainer<?> mysql = new MySQLContainer<>(MySQLTestImages.MYSQL_56_IMAGE)
+                .withConfigurationOverride("somepath/mysql_conf_override")
+                .withLogConsumer(new Slf4jLogConsumer(logger))
+        ) {
+            URL resource = this.getClass().getClassLoader().getResource("somepath/mysql_conf_override");
+
+            File file = new File(resource.toURI());
+            assertThat(file.isDirectory()).isTrue();
+
+            Set<PosixFilePermission> permissions = new HashSet<>(
+                Arrays.asList(
+                    PosixFilePermission.OWNER_READ,
+                    PosixFilePermission.OWNER_WRITE,
+                    PosixFilePermission.OWNER_EXECUTE
+                )
+            );
+
+            Files.setPosixFilePermissions(file.toPath(), permissions);
+
+            mysql.start();
+            assertThatCustomIniFileWasUsed(mysql);
+        }
+    }
+
     private void assertHasCorrectExposedAndLivenessCheckPorts(MySQLContainer<?> mysql) {
         assertThat(mysql.getExposedPorts()).containsExactly(MySQLContainer.MYSQL_PORT);
         assertThat(mysql.getLivenessCheckPortNumbers()).containsExactly(mysql.getMappedPort(MySQLContainer.MYSQL_PORT));
+    }
+
+    private void assertThatCustomIniFileWasUsed(MySQLContainer<?> mysql) throws SQLException {
+        try (ResultSet resultSet = performQuery(mysql, "SELECT @@GLOBAL.innodb_file_format")) {
+            String result = resultSet.getString(1);
+            assertThat(result).as("The InnoDB file format has been set by the ini file content").isEqualTo("Barracuda");
+        }
     }
 }

